server/keyshare/tasks/tasks.go

```diff
@@ -115,40 +140,65 @@ func (t *taskHandler) sendExpiryEmails(ctx context.Context, id int64, username,
 				email,
 				lang,
 			)
-			return err
+
+			if t.revalidateMail && err != nil && err != keyshare.ErrNoNetwork {
+				// When email revalidation is enabled and sending was impossible because of
+				// (temporary) MX / A record issues at the domain or an invalid email address,
+				// we mark the record to be revalidated in 5 days from now.
+				if err = t.db.ExecUserContext(ctx, "UPDATE irma.emails SET revalidate_on = $1 WHERE id = $2",
+					time.Now().AddDate(0, 0, 5).Unix(),
+					id); err != nil {
+					t.conf.Logger.WithField("error", err).Error("could not update email address to set revalidate_on")
+					return err
+				}
+				return nil
```

Comment:
If I'm right the situation might occur now that if a user has two email addresses of which the first is valid and the second one isn't, then the user will already receive mail on the first email address that the account is going to expire while the account is actually flagged for revalidation due to the second email address.

Reply:
You found something here, made a small tweak in sendExpiryEmails so a scenario where the user could receive multiple emails on the same address cannot happen anymore. Let me put it in writing to clarify:

Lets say:
User A has x (1, 2 or more) valid email addresses.
User B has 1 valid and 1 invalid email address.
User C has 1 invalid email address.
All users have more than 1 year of inactivity.

Now expireAccounts runs.

Without revalidation: 
- User A gets x expiry emails, account is flagged for deletion in 30d.
- User B gets 1 expiry email, the user 'hangs' forever inside expireAccounts making the effective LIMIT -=1.
- User C gets no expiry mail, the user 'hangs' forever inside expireAccounts making the effective LIMIT -=1.

With revalidation:
- User A gets x expiry mails, account is flagged for deletion in 30d.
- User B gets no expiry mail (yet). The invalid email address is flagged for revalidation. If the email address is still invalid after 5 days, it is deleted. If the email address is valid after 5 days, revalidate_on is removed from it. In both ways: with next run of expireAccounts now the user gets 1 or 2 expiry mails and account is flagged for deletion in 30d. 
- User C gets no expiry mail (yet). If the email address is still invalid after 5 days, it is deleted. The user now has no email addresses and stays (for now) forever in the db. If the email address is valid after 5 days, revalidate_on is removed and with the next run of expireAccounts the user receives 1 expiry mail and account is flagged for deletion in 30d.

Reply:
Sounds great. This is not implemented yet, right?

You should also check all derived cases of this issue type, like
https://github.com/privacybydesign/irmago/pull/313#discussion_r1241968456

Reply:
It is implemented, with the following addition now: on all relevant places first validation logic is applied to all users' email addresses. And then only the valid one's receive a mail (in case of multiple it's a BCC mail).